### ISSUE TITLE
validateBean - validation message per bean's property

### ISSUE THREAD

**Issue by ptrwis** [omnifaces/omnifaces#317]
Hello.
I have custom, class-level constraint validator (implementing `javax.validation.ConstraintValidator`) for a "Car" entity. This Car has a property called "color". The validator leaves a message per this property, like this:
```
context.buildConstraintViolationWithTemplate( message )
    .addPropertyNode( "color" ) //Car.color
    .addConstraintViolation();
```

Now, on my page, I have a form for creating a new Car, with selectOneMenu for choosing a "color":
```
<h:form>
    <h:outputText value="color" />
    <h:selectOneMenu value="#{createCar.car.color}" id="colorSelect" >
        <f:converter binding="#{viewScopedFacesConverter}" />
        <f:selectItems value="#{createCar.allColors}" var="color"   />
    </h:selectOneMenu>
    <h:message for="colorSelect" />
    <h:commandButton value="create" action="#{createCar.on_createCar_ButtonClick}" />
    <o:validateBean value="#{createCar.car}"  />
</h:form>
<h:messages  />
```
(other fields omitted for brevity, createCar is a controller)
As you can see, I'm using omnifaces/validateBean component to trigger my custom validator.
Everything works (almost) fine. The problem is that validateBean component leaves the message in `<h:messages>` (I believe in  fact it leaves a message with _form_ being the client), not in `<h:message for="colorSelect" />`, what I would expect.
Please consider adding support for leaving a message per property when validating whole bean.

Here is full source code of the validator:
```
public class CarModelVsColorValidatorImpl implements ConstraintValidator<CarModelVsColorValidator, Car>
{

    @Override
    public boolean isValid(Car car, ConstraintValidatorContext context) {
	if( car==null )
	    return true;
        boolean isValid = true; 
        
        String message="";
        
        // (...)
        if ( car.getModel().equals("BMW") && car.getColor().equals("BLUE") )
        {
            isValid = false;
            message = "Bmw can't be blue!";
        }
        
        if( isValid == false ){
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate( message )
                   .addPropertyNode( "color" )
                   .addConstraintViolation();
        }
        
        return isValid;
    }

    @Override
    public void initialize(CarModelVsColorValidator constraintAnnotation) {
        
    }
    
}
```

**Comment by BalusC** [comment]
I have added `showMessageFor` attribute with almost same logic as existing `ValidateMultipleFields` components (`<o:validateAll>`, `<o:validateEqual>`, etc).

```
<o:validateBean ... showMessageFor="colorSelect" />
```

Supported values are `@form` (default), `@all` (all associated inputs), `@global` (for `globalOnly="true"`) and any space separated string of (relative) client IDs.

**Comment by BalusC** [comment]
It's available in today's latest [2.6-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/org/omnifaces/omnifaces/2.6-SNAPSHOT/). Let me know if that does the job for you.

**Comment by ptrwis** [comment]
Well... no it's not working. I'm trying latest 2.6-20160926.134838-12. The message is still showing under `<h:messages>`. But I'm not sure if `showMessageFor` attribute is a good idea- what if `ConstraintValidator` throw many `ConstraintViolations`, or there are many class-level `ConstraintValidators` on one entity?

**Comment by BalusC** [comment]
> Well... no it's not working. I'm trying latest 2.6-20160926.134838-12. The message is still showing under `<h:messages>`.

This statement is ambiguous. Does it show up in `<h:message for="colorSelect" />` or not? The `<h:messages>` will by default display all messages, also the already displayed ones (which you can turn off with `redisplay="false"`). At least, it works for me and it's also covered in IT: https://github.com/omnifaces/omnifaces/commit/99e58035d2a463b7f8e01e562cfccb351d1d3518 

> But I'm not sure if `showMessageFor` attribute is a good idea- what if `ConstraintValidator` throw many `ConstraintViolations`, or there are many class-level `ConstraintValidators` on one entity?

You can also use a `<h:messages for="colorSelect">`. 

After re-reading your question I guess you'd like to associate the message with the input component bound to the property as specified by `addPropertyNode()`. But the `ConstraintViolation` has only a `getPropertyPath()` method which doesn't return expected values.

**Comment by ptrwis** [comment]
> After re-reading your question I guess you'd like to associate the message with the input component bound to the property as specified by `addPropertyNode()`. But the `ConstraintViolation` has only a `getPropertyPath()` method which doesn't return expected values.

Yes this is what I wanted to do. `ConstraintViolation` has also methods like `getLeafBean` or `getRootBean`. They reference specific bean's instance, the same as `ValueExpression.getValue()` does. I thought that maybe somehow combining `getRootBean` with `getPropertyPath` and comparing with properties returned by `ExpressionLanguage` (`ValueExpression` etc) it could be possible to do what I was asking for. Now I think that maybe there should be another phase in JSF like "validate model", because it's obvious (?) that this (second) validation should be done after "_update model values_" phase.

**Comment by ptrwis** [comment]
I was trying to solve it by myself but I'm not sure if this is doable. For now I'll close the issue. Thank you for your help anyway :)

**Comment by recentralized** [comment]
FWIW I implemented the requested feature in #343. 

> After re-reading your question I guess you'd like to associate the message with the input component bound to the property as specified by addPropertyNode(). But the ConstraintViolation has only a getPropertyPath() method which doesn't return expected values.

Quite true, my implementation naïvely expects `Path#toString()` to return something like 'color' (which it does with Hibernate), and doesn't make any effort to navigate more complex/nested Path instances.  However the fall-through of treating unidentified ConstraintViolation→UIComponent mappings as if `"@form"` were specified works well enough.
